include/pybind11/embed.h

```diff
@@ -87,29 +87,126 @@ struct embedded_module {
     }
 };
 
+/// Python 2.x/3.x-compatible version of `PySys_SetArgv`
+inline void set_interpreter_argv(int argc, char** argv, bool add_current_dir_to_path) {
+    // Before it was special-cased in python 3.8, passing an empty or null argv
+    // caused a segfault, so we have to reimplement the special case ourselves.
+    char** safe_argv = argv;
+    if (nullptr == argv || argc <= 0) {
+        safe_argv = new char*[1];
+        if (nullptr == safe_argv) return;
+        safe_argv[0] = new char[1];
+        if (nullptr == safe_argv[0]) {
+            delete[] safe_argv;
+            return;
+        }
+        safe_argv[0][0] = '\0';
+        argc = 1;
+    }
+#if PY_MAJOR_VERSION >= 3
+    // SetArgv* on python 3 takes wchar_t, so we have to convert.
+    wchar_t** widened_argv = new wchar_t*[static_cast<unsigned>(argc)];
+    for (int ii = 0; ii < argc; ++ii) {
+#  if PY_MINOR_VERSION >= 5
+        // From Python 3.5 onwards, we're supposed to use Py_DecodeLocale to
+        // generate the wchar_t version of argv.
+        widened_argv[ii] = Py_DecodeLocale(safe_argv[ii], nullptr);
```

Comment:
Ah, but that would make the loop condition a signed comparison error

Reply:
Do you think it would be cleaner if I did

```cpp
size_t argc_unsigned = static_cast<unsigned>(argc)
```

at the top of the function and then just used `size_t`s throughout?

Reply:
:facepalm: ...nope. If I do that I get narrowing conversion errors on the PySys_SetArgv calls.

Reply:
This is one of the impossible things to get completely right without using some casts. The array indices are `size_t`, yet `argv` is an `int`.

I'm not going to nag about this, but there is some inconsistency in this patch.

Reply:
Improved slightly (`ii` is now `size_t`) but still a little awkward.